File: ForwardList.java
package structures;

import exceptions.IndexOutOfRangeException;

public class ForwardList extends Structure {

    // Убрали static, если нужно использовать в методах removeBefore/removeAfter
    private class Node {
        String data;
        Node next;

        Node(String data) {
            this.data = data;
            this.next = null;
        }
    }
    
    private Node head;
    private Node tail;
    private int size;

    public ForwardList() {
        super();
        head = null;
        tail = null;
        size = 0;
    }

    public ForwardList(String name) {
        super(name);
        head = null;
        tail = null;
        size = 0;
    }

    public void pushFront(String value) {
        Node newNode = new Node(value);

        if (head == null) {
            head = newNode;
            tail = newNode;
        } else {
            newNode.next = head;
            head = newNode;
        }
        size++;
    }

    public void pushBack(String value) {
        Node newNode = new Node(value);

        if (tail == null) {
            head = newNode;
            tail = newNode;
        } else {
            tail.next = newNode;
            tail = newNode;
        }
        size++;
    }

    public void insertBefore(int position, String value) throws IndexOutOfRangeException {
        if (position < 0 || position > size) {
            throw new IndexOutOfRangeException("Position out of range: " + position);
        }

        if (position == 0) {
            pushFront(value);
        } else if (position == size) {
            pushBack(value);
        } else {
            Node prev = getNodeAt(position - 1);
            Node newNode = new Node(value);
            newNode.next = prev.next;
            prev.next = newNode;
            size++;
        }
    }

    public void insertAfter(int position, String value) throws IndexOutOfRangeException {
        if (position < 0 || position >= size) {
            throw new IndexOutOfRangeException("Position out of range: " + position);
        }

        if (position == size - 1) {
            pushBack(value);
        } else {
            Node current = getNodeAt(position);
            Node newNode = new Node(value);
            newNode.next = current.next;
            current.next = newNode;
            size++;
        }
    }

    public String popFront() throws IndexOutOfRangeException {
        if (head == null) {
            throw new IndexOutOfRangeException("List is empty");
        }

        String value = head.data;
        head = head.next;

        if (head == null) {
            tail = null;
        }

        size--;
        return value;
    }

    public String popBack() throws IndexOutOfRangeException {
        if (tail == null) {
            throw new IndexOutOfRangeException("List is empty");
        }

        if (head == tail) {
            String value = head.data;
            head = null;
            tail = null;
            size--;
            return value;
        }

        Node current = head;
        while (current.next != tail) {
            current = current.next;
        }

        String value = tail.data;
        current.next = null;
        tail = current;
        size--;

        return value;
    }

    public String get(int position) throws IndexOutOfRangeException {
        return getNodeAt(position).data;
    }

    private Node getNodeAt(int position) throws IndexOutOfRangeException {
        if (position < 0 || position >= size) {
            throw new IndexOutOfRangeException("Position out of range: " + position);
        }

        Node current = head;
        for (int i = 0; i < position; i++) {
            current = current.next;
        }

        return current;
    }

    public void removeBefore(Node targetNode) throws IndexOutOfRangeException {
        if (targetNode == null) {
            throw new IndexOutOfRangeException("Target node is null");
        }

        if (head == null || targetNode == head) {
            throw new IndexOutOfRangeException("Cannot remove before head");
        }

        if (head.next == targetNode) {
            popFront();
            return;
        }

        Node current = head;
        while (current != null && current.next != null && current.next.next != targetNode) {
            current = current.next;
        }

        if (current == null || current.next == null) {
            throw new IndexOutOfRangeException("Cannot find element to remove");
        }

        Node toDelete = current.next;
        current.next = targetNode;
        size--;
    }

    public void removeAfter(Node prevNode) throws IndexOutOfRangeException {
        if (prevNode == null || prevNode.next == null) {
            throw new IndexOutOfRangeException("Node has no next element");
        }

        Node toDelete = prevNode.next;
        prevNode.next = toDelete.next;

        if (toDelete == tail) {
            tail = prevNode;
        }

        size--;
    }

    public Node findByValue(String value) {
        Node current = head;
        while (current != null) {
            if (current.data.equals(value)) {
                return current;
            }
            current = current.next;
        }
        return null;
    }

    public int getSize() {
        return size;
    }

    @Override
    public String serialize() {
        StringBuilder sb = new StringBuilder();
        sb.append("F ").append(name).append(" ").append(size);

        Node current = head;
        while (current != null) {
            sb.append(" ").append(current.data);
            current = current.next;
        }

        return sb.toString();
    }

    @Override
    public void deserialize(String data) {
        String[] parts = data.split(" ");
        if (parts.length < 3) {
            return;
        }

        this.name = parts[1];
        int count = Integer.parseInt(parts[2]);

        // Clear existing list
        head = null;
        tail = null;
        size = 0;

        for (int i = 0; i < count && (i + 3) < parts.length; i++) {
            pushBack(parts[i + 3]);
        }
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("ForwardList \"").append(name).append("\" (size: ").append(size).append("): [");

        Node current = head;
        while (current != null) {
            if (current != head) {
                sb.append(", ");
            }
            sb.append(current.data);
            current = current.next;
        }

        sb.append("]");
        return sb.toString();
    }
}

File: forward_list.rs
// src/forward_list.rs
use crate::Structure;

#[derive(Debug)]
pub struct FNode {
    pub key: String,
    pub next: Option<Box<FNode>>,
}

#[derive(Debug)]
pub struct ForwardList {
    pub head: Option<Box<FNode>>,
    pub size: usize,
    pub name: String,
}

impl ForwardList {
    pub fn new() -> Self {
        ForwardList {
            head: None,
            size: 0,
            name: String::new(),
        }
    }
    
    pub fn push_front(&mut self, key: String) {
        let new_node = Box::new(FNode {
            key,
            next: self.head.take(),
        });
        self.head = Some(new_node);
        self.size += 1;
    }
    
    pub fn push_back(&mut self, key: String) {
        let new_node = Box::new(FNode { key, next: None });
        
        if let Some(ref mut head) = self.head {
            let mut current = head;
            while let Some(ref mut next) = current.next {
                current = next;
            }
            current.next = Some(new_node);
        } else {
            self.head = Some(new_node);
        }
        self.size += 1;
    }
    
    pub fn pop_front(&mut self) -> Option<String> {
        self.head.take().map(|node| {
            self.head = node.next;
            self.size -= 1;
            node.key
        })
    }
    
    pub fn front(&self) -> Option<&str> {
        self.head.as_ref().map(|node| node.key.as_str())
    }
    
    pub fn clear(&mut self) {
        self.head = None;
        self.size = 0;
    }
    
    pub fn to_vec(&self) -> Vec<String> {
        let mut result = Vec::new();
        let mut current = self.head.as_ref();
        
        while let Some(node) = current {
            result.push(node.key.clone());
            current = node.next.as_ref();
        }
        
        result
    }
    
    pub fn contains(&self, value: &str) -> bool {
        let mut current = self.head.as_ref();
        
        while let Some(node) = current {
            if node.key == value {
                return true;
            }
            current = node.next.as_ref();
        }
        
        false
    }
    
    pub fn len(&self) -> usize {
        self.size
    }
    
    pub fn is_empty(&self) -> bool {
        self.size == 0
    }
    
    pub fn get(&self, index: usize) -> Option<&str> {
        if index >= self.size {
            return None;
        }
        
        let mut current = self.head.as_ref();
        for _ in 0..index {
            current = current.and_then(|node| node.next.as_ref());
        }
        
        current.map(|node| node.key.as_str())
    }
}

impl Structure for ForwardList {
    fn serialize(&self) -> String {
        let values: Vec<String> = self.to_vec();
        format!("F {} {} {}", self.name, self.size, values.join(" "))
    }
    
    fn deserialize(&mut self, data: &str) {
        let parts: Vec<&str> = data.split_whitespace().collect();
        if parts.len() < 3 || parts[0] != "F" {
            return;
        }
        
        self.name = parts[1].to_string();
        let count: usize = parts[2].parse().unwrap_or(0);
        
        self.clear();
        
        for i in 0..count {
            if let Some(value) = parts.get(3 + i) {
                self.push_back(value.to_string());
            }
        }
    }
    
    fn name(&self) -> &str {
        &self.name
    }
    
    fn set_name(&mut self, name: String) {
        self.name = name;
    }
}
